mongodbatlas/resource_mongodbatlas_cluster_migrate.go

```diff
@@ -0,0 +1,368 @@
+				Computed: true,
+			},
+			"paused": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"srv_address": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"state_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"advanced_configuration": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"fail_index_key_too_long": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Computed: true,
+						},
+						"javascript_enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Computed: true,
+						},
+						"minimum_enabled_tls_protocol": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+						},
+						"no_table_scan": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Computed: true,
+						},
+						"oplog_size_mb": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Computed: true,
+						},
+						"sample_size_bi_connector": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Computed: true,
+						},
+						"sample_refresh_interval_bi_connector": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Computed: true,
+						},
+					},
+				},
+			},
+			"labels": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Set: func(v interface{}) int {
+					var buf bytes.Buffer
+					m := v.(map[string]interface{})
+					buf.WriteString(m["key"].(string))
+					buf.WriteString(m["value"].(string))
+					return hashcode.String(buf.String())
+				},
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+						},
+						"value": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+						},
+					},
+				},
+			},
+			"snapshot_backup_policy": computedCloudProviderSnapshotBackupPolicySchema(),
+		},
+	}
+}
+
+func resourceMongoDBAtlasClusterStateUpgradeV0(rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	log.Println("[INFO] Found MongoDB Cluser state v0; migrating to v1")
+	rawState = migrateAdvancedConfiguration(rawState)
+
+	return rawState, nil
+}
+
+func migrateAdvancedConfiguration(rawState map[string]interface{}) map[string]interface{} {
+	if v, ok := rawState["advanced_configuration"]; ok || v != nil {
+		advancedConfig := v.(map[string]interface{})
+		if len(advancedConfig) != 0 {
+			if advancedConfig["oplog_size_mb"] == "" {
+				advancedConfig["oplog_size_mb"] = "0"
+			}
+			if advancedConfig["sample_refresh_interval_bi_connector"] == "" {
+				advancedConfig["sample_refresh_interval_bi_connector"] = "0"
+			}
+			if advancedConfig["sample_size_bi_connector"] == "" {
+				advancedConfig["sample_size_bi_connector"] = "0"
+			}
```

Comment:
@gmlp Could you explain what this part of the migration is doing? This look potentially dangerous.

Reply:
@shum, This part is for clusters that were created using v0.5.1 without advanced_config argument. The state will be equal to  https://github.com/terraform-providers/terraform-provider-mongodbatlas/commit/4259166edaabd005fb60165a7fe2fa17df1dea48#diff-92bb51ed35b9397957bd08d5cd49ed0bR119-R121 in order to make it compatible with the new schema. we need to parse this empty string to 0.

Reply:
@marinsalinas , @PacoDw any thoughts here guys?

Reply:
If I'm understanding correctly however, this means that any cluster created without advanced_config args would now be migrated to have oplog size of 0 because it was previously an empty string? I think this is potentially incorrect.

Reply:
So we're fetching `advanced_configuration` from `rawState`, then if `oplog_size_mb` wasn't explicitly set by the user ie `""`, then we default it to 0? Is there a way where we continue to leave it out and just assume Atlas defaults if it's not provided?

Reply:
@shum this is such a tricky one - when one creates a cluster in Atlas you can't send the advanced configs at the same time to the API.  They are only modifiable AFTER the cluster is created (https://docs.atlas.mongodb.com/reference/api/clusters-modify-advanced-configuration-options/).  So if we want users to be able to configure both the cluster and the advanced configuration options at the same time (like they can in the UI) via Terraform it's tricky.   The best way would likely be to create the cluster and then populate the state after the cluster is created by pulling the defaults from Atlas (https://docs.atlas.mongodb.com/reference/api/clusters-get-advanced-configuration-options/) but if the users doesn't have these defined in their TF file it's then something in state that doesn't really match the tf config. We sorta did this for the snapshot backup policy and it's a bit awkward due to that (because it always sets a default policy but again we have a total mismatch from the UI so it's confusing - we have an issue on it actually).   

This is why I wanted your review here.  I'm very worried we will inadvertently cause an issue due to the complexity of not really being having the full CRUD life-cycle available.   Perhaps we should do a test where a user makes a change to a cluster where these are set to zero and see if the terraform plan sees it as a change to apply.  That would signal a big issue.

Reply:
Hi @themantissa,@shum, I've been doing some research and testing. I figure out that this can improve and I can skip these conditions. I'll open a new PR with the improvement.

Reply:
@gmlp thank you so much! Will keep an eye out.

Reply:
@gmlp Let me know how you want to proceed with this PR and if we're excluding the migration step from this PR.

Reply:
We can't just exclude the conversion between empty string and `0` otherwise we will get the `Error: a number is required` again. See the messages on this PR where that's described. Unless this code is removed and something else is changed together.

Reply:
Can't we change this something like this to avoid unwanted empty values?


```go
func migrateAdvancedConfiguration(rawState map[string]interface{}) map[string]interface{} {
	if v, ok := rawState["advanced_configuration"]; ok || v != nil {
		advancedConfig := v.(map[string]interface{})
		if len(advancedConfig) != 0 {

			advaceConfigMap := map[string]interface{}{
				"fail_index_key_too_long":              advancedConfig["fail_index_key_too_long"],
				"javascript_enabled":                   advancedConfig["javascript_enabled"],
				"minimum_enabled_tls_protocol":         advancedConfig["minimum_enabled_tls_protocol"],
				"no_table_scan":                        advancedConfig["no_table_scan"],
			},

			if v1, ok := advancedConfig["oplog_size_mb"]; ok, v1!= "" {
				advaceConfigMap["oplog_size_mb"] = v["oplog_size_mb"].(string)
			}
			if v1, ok := advancedConfig["sample_refresh_interval_bi_connector"]; ok, v1!= ""  {
				advaceConfigMap["sample_refresh_interval_bi_connector"] =  v["sample_refresh_interval_bi_connector"].(string)
			}
			if v1, ok := advancedConfig["sample_size_bi_connector"]; ok, v1!= ""  {
				advaceConfigMap["sample_size_bi_connector"] =  v["sample_size_bi_connector"].(string)
			}

			rawState["advanced_configuration"] = []interface{}{
				advaceConfigMap,
			}

		} else {
			rawState["advanced_configuration"] = []interface{}{}
		}
	} else {
		rawState["advanced_configuration"] = []interface{}{}
	}
	return rawState
}
```

Reply:
Thanks @marinsalinas, agree that something like this should work as well. Let me know how you and @gmlp want to proceed.

Reply:
Hi @themantissa, @shum the new pull request with option 2 is: https://github.com/terraform-providers/terraform-provider-mongodbatlas/pull/238